// Copyright (c) YugaByte, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied.  See the License for the specific language governing permissions and limitations
// under the License.
//

syntax = "proto3";

import "google/protobuf/descriptor.proto";

package yb.rpc;

extend google.protobuf.MethodOptions {
  LightweightMethodOptions lightweight_method = 50000;
}

enum RpcSides {
  NONE = 0;
  SERVICE = 1;
  PROXY = 2;
  BOTH = 3;
}

message LightweightMethodOptions {
  RpcSides sides = 1;
}

extend google.protobuf.FieldOptions {
  LightweightFieldOptions lightweight_field = 50000;
}

message LightweightFieldOptions {
  bool pointer = 1;
}

extend google.protobuf.MessageOptions {
  LightweightMessageOptions lightweight_message = 50000;
}

message LightweightMessageOptions {
  bool force_arena = 1;
  bool verify = 2;
  bool dump_parse = 3;
}

extend google.protobuf.EnumOptions {
  LightweightEnumOptions lightweight_enum = 50000;
}

message LightweightEnumOptions {
  string name = 1;
}
